### PR TITLE
fix: remote write should sort labels before pushing

### PR DESCRIPTION
**Pull request by 123BLiN**
When we try to push metrics to our Thanos setup, we got an error:
```
component=receive component=receive-writer tenant=default-tenant msg="Out of order labels in the label set" 
```
Quick googling gives a couple of known issues:
- https://github.com/thanos-io/thanos/issues/6452
- https://github.com/prometheus/prometheus/issues/11505

So this PR tries to sort labels before pushing to remote write

~Trying to test it locally somehow~
We have tested in our EKS setup, with the provided fix we can see the metrics in Thanos now
Screenshot 2024-05-23 at 14 19 06

We now get another warnings like:
```
thanos-tenants-receive-0 receive level=debug ts=2024-05-23T12:28:44.029456659Z caller=writer.go:91 component=receive component=receive-writer tenant=default-tenant msg="Labels with empty name in the label set" lset="labels:<name:\"__name__\" value:\"egressd_transmit_bytes_total\" > labels:<name:\"cross_zone\" value:\"false\" > labels:<name:\"dst_ip\" value:\"10.10.181.95\" > labels:<name:\"dst_ip_type\" value:\"private\" > labels:<name:\"dst_namespace\" > labels:<name:\"dst_node\" > labels:<name:\"dst_pod\" > labels:<name:\"dst_zone\" > labels:<name:\"proto\" value:\"TCP\" > labels:<name:\"src_ip\" value:\"10.10.157.101\" > labels:<name:\"src_namespace\" value:\"de\" > labels:<name:\"src_node\" value:\"ip-10-10-155-97.ec2.internal\" > labels:<name:\"src_pod\" value:\"airbyte-pod-sweeper-75cb44885b-yz246\" > labels:<name:\"src_zone\" value:\"us-east-1a\" > "
```

which is probably expected since no destination pod was found in the cluster
but that probably means we should omit these metrics instead of sending them empty or set them to some static value like 0,
~we will try to come up with another fix for that~
UPD: I've added a small opinionated fix - replace empty values with `"unknown"`
Now it works for us without errors

P.P.S.
if anyone wants to test this PR with his deployment, here are the steps we've come up with:
- build and push to your registry
```
GOOS=linux GOARCH=amd64 go build -ldflags="-s -w" -o bin/egressd-exporter-x86_64 cmd/exporter/main.go

docker build --build-arg TARGETARCH=x86_64 -f Dockerfile.exporter -t 1234567890.dkr.ecr.us-east-1.amazonaws.com/egressd-exporter:latest .

docker push 1234567890.dkr.ecr.us-east-1.amazonaws.com/egressd-exporter:latest
```

- replace the image in your deployment 
- config we run with:
```
egressd:
  exporter:
    config: |
      exportInterval: 15s
      sinks:
        prom_sink:
          prom_remote_write:
            url: "thanos-tenants-receive.monitoring.svc:19291/api/v1/receive"
```

### CODE REVIEW

File: exporter/sinks/prom.go [hunk 1]
```diff
@@ -50,25 +51,29 @@ func (s *PromRemoteWriteSink) Push(ctx context.Context, batch *pb.PodNetworkMetr
...
+			{Name: "dst_zone", Value: ensureNonEmpty(m.DstZone)},
+			{Name: "dst_ip", Value: dstIP.String()},
+			{Name: "dst_ip_type", Value: dstIPType},
+			{Name: "cross_zone", Value: isCrossZoneValue(m)},
+
+			{Name: "proto", Value: protoString(uint8(m.Proto))},
+		}
+		sort.Slice(labels, func(i, j int) bool {
+			return labels[i].Name < labels[j].Name
+		})
```
Comment (artem-ag) [inline exporter/sinks/prom.go:74]: @123BLiN From pure performance point of view consider using pre-sorted labels here (as [you already did in a test](https://github.com/castai/egressd/pull/109/files#diff-4ee7f1b77c14090b7b7b96fd5574a1e59fe437410fd958a8c8ec2bc31ad20708R63-R76))
And in prom_test you can take this Slice, sort it, and verify that it did not change after sorting.

**Comment by ehaelyon** [comment]
Hi,
Any idea when this PR will be merged? 
This fix is very important for our organization and we are looking forward to it :)

**Comment by anjmao** [comment]
Hi, I will have a look at PR soon and merge it.

**Comment by anjmao** [comment]
Fixed with https://github.com/castai/egressd/commit/9bbb75e01b08996123a84cf84a7d4372a57049c2

**Pull request closed without merging**